File: week1_degrees/degrees.rb
#!/usr/bin/env ruby
#
# Degree Conversions - Ruby Solution
#
# To Run: `ruby degrees.rb` or `./degrees.rb`
#

# Functions to convert degrees <-> radians ======
# Note: Parentheses are optional in Ruby
def degrees_to_radians degrees
    # The :: operator is much like the scope
    # resolution operator in C++
    degrees * Math::PI / 180.0
end

def radians_to_degrees radians
    # Also note how in Ruby the last line of a
    # function is the value the function returns.
    # the 'return' keyword is optional in this case.
    radians * 180.0 / Math::PI
end

# Functions to convert fahrenheit <-> celsius <-> kelvins ========

def fahrenheit_to_celsius f
    5.0 / 9.0 * (f - 32.0)
end

def celsius_to_fahrenheit c
    9.0 / 5.0 * c + 32.0
end

def celsius_to_kelvin c
    c + 273.15
end

def kelvin_to_celsius k
    k - 273.15
end

# Programming Puzzle Functions

def read_input
    # gets -- reads a line of input
    # strip -- removes whitespace from beginning
    #   and end of the string
    input_line = gets.strip
    output_line = gets.strip

    # Return the values as an array
    [input_line, output_line]
end

def parse_input_line line
    # Split the lines into three parts. value needs to
    # be a float (decimal number).
    #
    # [0...-2] means slice the text from the start (position 0)
    #   to 2 from the right (exclusive).
    # Example: "hello"[0...-2] = "hel"
    #
    # .to_f turns a string (text) into a float (decimal)
    value = line[0...-2].to_f
    # Select the last two characters. Double dot means include
    # index -1 where triple dot would mean exclude -1
    units = line[-2..-1]

    # We want to return the value and the units
    [value, units]
end

def convert value, units
    # Check the value of units and return one
    # of the following results. This is like
    # a switch statement in other languages but
    # a little nicer
    case units
    when "rd"
        # Parentheses are optional!
        radians_to_degrees value
    when "dr"
        degrees_to_radians value
    when "fc"
        fahrenheit_to_celsius value
    when "cf"
        celsius_to_fahrenheit value
    when "ck"
        celsius_to_kelvin value
    when "kc"
        kelvin_to_celsius value
    when "fk"
        # You can compose functions in programming!
        celsius_to_kelvin(fahrenheit_to_celsius(value))
    when "kf"
        # If you have to nest function calls, keep the
        # parentheses
        celsius_to_fahrenheit(kelvin_to_celsius(value))
    # The user typed something that isn't a valid conversion
    else
        # We can't convert this, so let's raise an error
        # and let someone else handle it.
        raise "No candidate for conversion"
    end
end

def process_line line
    # Reformat the input line
    value, units = parse_input_line line
    # Calculate the result
    result = convert value, units
    # Get the result's units (last character in the units string)
    result_units = units[-1]

    # Print the value rounded to two decimal places
    # and the result units
    puts "#{result.round(2)}#{result_units}"
end

# Main script
# Read the two lines of input. Note that we are
# implicitly calling the read_input function just by
# writing its name!
input_lines = read_input

# Loop over the input lines
# this is like `for line in input_lines` in Python
input_lines.each do |line|
    # This is like a try/catch block
    # in other languages
    begin
        process_line line
    rescue => error
        # If an error happens, print the error message
        puts error
    end
end

File: week1_degrees/degrees.py
#!/usr/bin/env python3
"""
Degrees Puzzle -- Python Solution

Author: Peter Gagliardi

To run: `python3 degrees.py` or `./degrees.py`
"""
# I'm a connoseiur of pi. my slices are imported.
import math

# Functions for converting between degrees <-> radians
def degrees_to_radians(degrees):
    return degrees * math.pi / 180.0

def radians_to_degrees(radians):
    return radians * 180.0 / math.pi

# Functions for converting °F <-> °C <-> K

def fahrenheit_to_celsius(f):
    return 5.0 / 9.0 * (f - 32.0)

def celsius_to_fahrenheit(c):
    return 9.0 / 5.0 * c + 32.0

def celsius_to_kelvin(c):
    return c + 273.15

def kelvin_to_celsius(k):
    return k - 273.15

# Programming challenge solution ========

def read_input():
    """
    Read input from the command line
    input -- read a line of input
    strip -- remove whitespace from beginning
        and end of string.
    """
    input_line = input().strip()
    output_line = input().strip()

    # Return the values as an array
    return input_line, output_line

def parse_input_line(line):
    # Python allows multiple variable assignment!
    # line[:-2] means slice from beginning of string up to
    # but not including the 2nd character from the right
    # line[-2:] means slice from the 2nd to last character to
    # the end
    value, units = float(line[:-2]), line[-2:]
    return value, units

def convert(value, units):
    if units == "rd":
        return radians_to_degrees(value)
    elif units == "dr":
        return degrees_to_radians(value)
    elif units == "fc":
        return fahrenheit_to_celsius(value)
    elif units == "cf":
        return celsius_to_fahrenheit(value)
    elif units == "ck":
        return celsius_to_kelvin(value)
    elif units == "kc":
        return kelvin_to_celsius(value)
    elif units == "kf":
        return celsius_to_fahrenheit(kelvin_to_celsius(value))
    elif units == "fk":
        return celsius_to_kelvin(fahrenheit_to_celsius(value))
    else:
        raise ValueError("No candidate for conversion")

def process_line(line):
    """
    handle a single line of input
    """
    # Reformat the input line
    value, units = parse_input_line(line)
    # Calculate the result
    result = convert(value, units)
    # Get the result's units (last character in the units string)
    result_units = units[-1]

    # Print the value rounded to two decimal places
    # and the result units
    print("{:.02f}{}".format(result, result_units))

# This is a common Python idiom for
# making sure this code only runs when run as a script
# and not when imported as a library
if __name__ == "__main__":
    # Read the two lines of input.
    input_lines = read_input()

    # Loop over both lines
    for line in input_lines:
        try:
            process_line(line)
        # If something goes wrong, print the error
        # message and keep going.
        except ValueError as e:
            print(e)
